src/abstractdataframe/selection.jl

```diff
@@ -0,0 +1,357 @@
+In particular, regular expressions, `All`, `Between`, and `Not` selectors are supported.
+
+Column renaming and transformations are supported.
+The syntax for column renaming is `old_column=>new_column_name`.
+The syntax for column transformations is `old_column=>fun=>new_column_name`.
+`old_column` can be a `Symbol` or an integer, ``new_column_name` must be a `Symbol`,
+and `fun` must be a `Function` that is applied row by row to the values of `old_column`.
+
+Column transfomation syntax also supports a short `old_column=>fun` form, in which
+case `new_column_name` is automatically generated by joining `olcd_column` with `fun` name
+with `_`.
+
+If more than one argument is passed then duplicates are accepted except for
+column renaming and transformation operations, where it is not alloweded to rename/transform
+into the same column name.
+For example if `:col` is present in `df` a call `select!(df, :col, :)` is valid
+and moves the column `:col` moved to be the first one in-place.
+
+Note that including the same column several times in the data frame will create aliases.
+
+# Examples
+```jldoctest
+julia> df = DataFrame(a=1:3, b=4:6)
+3×2 DataFrame
+│ Row │ a     │ b     │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 4     │
+│ 2   │ 2     │ 5     │
+│ 3   │ 3     │ 6     │
+
+julia> select!(df, 2)
+3×1 DataFrame
+│ Row │ b     │
+│     │ Int64 │
+├─────┼───────┤
+│ 1   │ 4     │
+│ 2   │ 5     │
+│ 3   │ 6     │
+
+julia> df = DataFrame(a=1:3, b=4:6)
+3×2 DataFrame
+│ Row │ a     │ b     │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 4     │
+│ 2   │ 2     │ 5     │
+│ 3   │ 3     │ 6     │
+
+julia> select!(df, :a=><(1.5)=>:c, :b)
+3×2 DataFrame
+│ Row │ c    │ b     │
+│     │ Bool │ Int64 │
+├─────┼──────┼───────┤
+│ 1   │ 1    │ 4     │
+│ 2   │ 0    │ 5     │
+│ 3   │ 0    │ 6     │
+```
+
+"""
+function select!(df::DataFrame, inds::AbstractVector{Int})
+    if isempty(inds)
+        empty!(_columns(df))
+        empty!(index(df))
+        return df
+    end
+    indmin, indmax = extrema(inds)
+    if indmin < 1
+        throw(ArgumentError("indices must be positive"))
+    end
+    if indmax > ncol(df)
+        throw(ArgumentError("indices must not be greater than number of columns"))
+    end
+    if !allunique(inds)
+        throw(ArgumentError("indices must not contain duplicates"))
+    end
+    copy!(_columns(df), _columns(df)[inds])
+    x = index(df)
+    copy!(_names(x), _names(df)[inds])
+    empty!(x.lookup)
+    for (i, n) in enumerate(x.names)
+        x.lookup[n] = i
+    end
+    return df
+end
+
+select!(df::DataFrame, c::Int) = select!(df, [c])
+select!(df::DataFrame, c::Union{AbstractVector{<:Integer}, AbstractVector{Symbol},
+                                Colon, All, Not, Between, Regex}) =
+    select!(df, index(df)[c])
+
+function select!(df::DataFrame, cs...)
+    newdf = select(df, cs..., copycols=false)
```

Comment:
Isn't this backwards? In general we define `f(x)` by calling `f!(copy(x))`.

Reply:
Not in this case as we have to avoid touching `df` when doing transformations (old columns in `df` might get updated in the process), so even in `select!` we have to remember the old `df` as-is. This is needed to make sure that e.g. the following code works correctly:
```
julia> df
1×2 DataFrame
│ Row │ a     │ b     │
│     │ Int64 │ Int64 │
├─────┼───────┼───────┤
│ 1   │ 1     │ 2     │

julia> select!(df, :a=>:b, :b=>:a)
1×2 DataFrame
│ Row │ b     │ a     │
│     │ Int64 │ Int64 │
├─────┼───────┼───────┤
│ 1   │ 1     │ 2     │
```
and if we need it then making a `copy` of `df` would be an extra unnecessary cost.